### DISCUSSION TITLE
Can I just override some styles of the default theme?

### DISCUSSION THREAD

**Discussion by Nefcanto** [keycloak/keycloak#20483] [Q&A]
To create a new theme, we need to create a file called `theme.properties`.

If in that file we don't include the `style` keyword, then the default keycloak theme would be used.

However, if we use that `style` then we should start from scratch because get would get zero default style.

It's like a binary, zero/one state. Is there a way to get something in between?

Can I use the default theme's style for my custom theme, and only override some CSS styles? (coloring for example).

**Comment by danielFesenmeyer** [comment]
What we do is that we add all the required styles and scripts from the parent themes to our custom theme and add our custom styles and scripts additionally. Like that:

```
styles=css/parent_1.css css/parent_2.css css/custom_1.css css/custom_2.css
scripts=js/parent.js js/custom.js
```

I don't think there is another way currently, but I'm not 100% sure.

**Comment by Nefcanto** [comment]
@danielFesenmeyer thank you. That works for me. But how do I get all of the styles and scripts of the parent theme? Should I find them by chance?

**Comment by danielFesenmeyer** [comment]
@Nefcanto You can just look what's the defined in the parent theme's theme properties. For example, for the keycloak login theme, you can find the settings here: https://github.com/keycloak/keycloak/blob/e41e1a971ae5b14d9ffd0292544f0073fa7f3da5/themes/src/main/resources/theme/keycloak/login/theme.properties#L4

(seems there are only styles, no scripts).

**Comment by Nefcanto** [comment]
@danielFesenmeyer, I got it. One more question and it's solved, how should I include it in my custom theme? I mean `style=css/login.css` is for the `keycloak` theme, but in my `customTheme` what should I write? Would it be like `style=css/login.css css/custom.css`? Does that result in 404 for `login.css`?

**Reply by danielFesenmeyer** [reply]
It should work as you described.

**Reply by Nefcanto** [reply]
I would test it. Thank you so much.

**Reply by Nefcanto** [reply]
@danielFesenmeyer No, that did not work. I added `styles=css/login.css css/styles.css` to my custom theme's `theme.properties` and as I expected,  the `loginc.ss` resulted in 404.

**Reply by danielFesenmeyer** [reply]
That's strange. I have a simple implemented a simple example login theme with Keycloak 20.0.1, and there it works perfectly fine. The `theme.properties` is just this:

```
# inherit everything from this parent, unless overwritten here
parent=keycloak

styles=css/login.css css/dtn-login.css

favIcon=/img/favicon.ico
```

Both `css/login.css` and my custom files are perfectly displayed.

**Reply by Nefcanto** [reply]
@danielFesenmeyer my bad. I forgot that `parent=keycloak` line. I added that and now it works perfectly. Both styles are loaded. You saved me from days of headache :D

**Comment by ssilvert** [comment]
We have created a quickstart for this.  It's not finished, so we haven't
announced it.  But it might help you get started.

https://github.com/keycloak/keycloak-quickstarts/tree/latest/extend-admin-console-quick-start

**Reply by Nefcanto** [reply]
@ssilvert one point, can you inject the current locale somewhere in the root? I want to customize Keycloak's theme, but I need to do some stuff for RTL languages. I know that [RTL](https://github.com/keycloak/keycloak/issues/17904) support is dependent on PatternFly's RTL support. But if you inject the current locale somewhere to help us write conditional CSS, it would help us a lot. For example:

```
body.ar #kc-form-login label {
    display: block !important;
    text-align: right;
}
```

**Reply by Nefcanto** [reply]
Also if you somehow give us the currently selected locale, we can change the realm display name dynamically to match the locale. For example, if the brand name is **Microsoft** in `en`, we would love to show **مایکروسافت** for `fa`. But there is only one field available for relam's display name. So it's not possible via the Admin Panel. But if we have the `en` or `fa` in the page, we can detect it and using JS change the realm name dynamically.

**Reply by ssilvert** [reply]
@Nefcanto We have started putting the locale in the `lang` attribute of the `html` tag.  Does that help you?

**Reply by Nefcanto** [reply]
@ssilvert yes of course. That's the standard approach. Thank you.

**Reply by Nefcanto** [reply]
@ssilvert I can't find it. Can you guide me please? In which version and where is the source code file?

**Reply by ssilvert** [reply]
Looks like I was mistaken about this.  The PR for it was abandoned due to inactivity.
https://github.com/keycloak/keycloak/pull/17112

**Reply by Nefcanto** [reply]
So, we're stuck for now. Is there any other way to do that? One way that comes to my mind is to check the URL parameter for `kc_locale` in our custom script.

**Reply by ssilvert** [reply]
We actually need to fix that PR and make it work properly.  Shouldn't be too hard.  I just don't know when I can get to it.  Maybe someone else will take it on?

@Nefcanto Are you modifying admin or account theme?  You could get the locale the same way we get it.

**Reply by jonkoops** [reply]
There is an issue to track this under #19965. If this is a priority for you feel free to send us a PR.

**Reply by Nefcanto** [reply]
@ssilvert I'm just working on the login pages (I guess that's account theme). @jonkoops I would love to, but I'm not a Java developer. I can't even run Keycloak on my local machine.

**Reply by jonkoops** [reply]
Understandable, we're working actively on customization so it will likely be prioritized (no promises). Right now we have some larger chunks to figure out first.

**Reply by Nefcanto** [reply]
@jonkoops thank you so much. We are using Keycloak for **all** of the applications we build for our customers and you guys have done an amazing job. I can't thank you enough for that.
